examples/vrf_on_shared_vc/main.go

```diff
@@ -0,0 +1,164 @@
+
+	for _, gateway := range gateways.MetalGateways {
+		api_client.MetalGatewaysApi.DeleteMetalGateway(ctx, gateway.MetalGateway.GetId()).Execute()
+	}
+
+	vrfs, r, err := api_client.VRFsApi.FindVrfs(ctx, projectId).Execute()
+
+	for _, vrf := range vrfs.Vrfs {
+		api_client.VRFsApi.DeleteVrf(ctx, vrf.GetId()).Execute()
+	}
+
+	vrf, r, err := api_client.VRFsApi.CreateVrf(ctx, projectId).VrfCreateInput(createVrfRequest).Execute()
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error when calling `VRFsApi.CreateVrf`: %v\n", err)
+		fmt.Fprintf(os.Stderr, "Full HTTP response: %v\n", r)
+		panic(err)
+	}
+
+	ipReservationRequest := metal.RequestIPReservationRequest{}
+	ipReservationRequest.VrfIpReservationCreateInput = &metal.VrfIpReservationCreateInput{
+		Type:    "vrf",
+		Cidr:    24,
+		Network: "10.0.0.0",
+		VrfId:   vrf.GetId(),
+	}
+
+	block, r, err := api_client.IPAddressesApi.RequestIPReservation(ctx, projectId).RequestIPReservationRequest(ipReservationRequest).Execute()
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error when calling `IPAddressesApi.RequestIPReservation`: %v\n", err)
+		fmt.Fprintf(os.Stderr, "Full HTTP response: %v\n", r)
+		panic(err)
+	}
+
+	createVlanRequest := metal.VirtualNetworkCreateInput{
+		Metro: &metro,
+	}
+
+	vlan, r, err := api_client.VLANsApi.CreateVirtualNetwork(ctx, projectId).VirtualNetworkCreateInput(createVlanRequest).Execute()
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error when calling `VLANsApi.CreateVirtualNetwork`: %v\n", err)
+		fmt.Fprintf(os.Stderr, "Full HTTP response: %v\n", r)
+		panic(err)
+	}
+
+	createGatewayRequest := metal.CreateMetalGatewayRequest{}
+	createGatewayRequest.VrfMetalGatewayCreateInput = &metal.VrfMetalGatewayCreateInput{
+		IpReservationId:  block.VrfIpReservation.GetId(),
+		VirtualNetworkId: vlan.GetId(),
+	}
+
+	_, r, err = api_client.MetalGatewaysApi.CreateMetalGateway(ctx, projectId).CreateMetalGatewayRequest(createGatewayRequest).Execute()
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error when calling `MetalGatewaysApi.CreateMetalGateway`: %v\n", err)
+		fmt.Fprintf(os.Stderr, "Full HTTP response: %v\n", r)
+		panic(err)
+	}
+
+	var speed int32 = 500000000
+	createConnectionRequest := metal.CreateOrganizationInterconnectionRequest{}
+	createConnectionRequest.VrfFabricVcCreateInput = &metal.VrfFabricVcCreateInput{
+		Metro:            metro,
+		Name:             "metal-go-example",
+		Redundancy:       "redundant",
+		ServiceTokenType: "a_side",
+		Type:             "shared",
+		Vrfs:             []string{vrf.GetId(), vrf.GetId()},
+		Speed:            &speed,
+	}
+
+	connection, r, err := api_client.InterconnectionsApi.CreateProjectInterconnection(ctx, projectId).CreateOrganizationInterconnectionRequest(createConnectionRequest).Execute()
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error when calling `InterconnectionsApi.CreateOrganizationInterconnection`: %v\n", err)
```

Comment:
BLARG.  Now the example is failing here:
```
Error when calling `InterconnectionsApi.CreateOrganizationInterconnection`: data matches more than one schema in oneOf(VirtualCircuit)
```
There are required fields defined on `VlanVirtualCircuit`, but those don't take effect right now (potential workaround for that generator issue is on #132).  In any case, required field validation wouldn't be enough here because `VrfVirtualCircuit` doesn't declare any required fields, and if it _did_ we'd have a problem of what to do when there aren't enough properties to match either of those (`Href` isn't an option in its current state because it allows properties other than `"href"`).

An alternative to required field validation, now that the SDK will return type mismatch errors, would be to add `vrf` with an invalid type (`integer`) to the `VlanVirtualCircuit` spec and to add `virtual_network` with an invalid type to the `VrfVirtualCircuit` spec.  But that approach has its own issues as well.

Reply:
I was trying to remember how I got this working without `required` support and finally got it.  The `MetalGateway` and `VrfMetalGateway` schemas can be disambiguated by leveraging the `type` field on IP reservations: a VRF IP reservation has one valid type, `vrf`, and a non-VRF IP reservation has multiple value types, none of which are `vrf`.  Since the type for each IP reservation schema is defined as an enum, `metal-go` is now capable of using that field to determine which schema to use; since a `MetalGateway` always has a non-VRF IP reservation and a `VrfMetalGateway` always has a VRF IP reservation, we can `include=ip_reservation` to help `metal-go` figure out the right gateway schema: https://github.com/equinix-labs/metal-go/pull/127/files#diff-49d8c1d06b64f8f0e02a49dd8b5f8efc1881471d3324bfbcf62e9f4403e311e0R86-R87

Reply:
This example has now been updated to use:

1. Invalid field specifications to differentiate VRF and VLAN virtual circuits (this allows us to disambiguate those schemas without implementing required field validation)
2. `include=ip_reservation` to disambiguate MetalGateway and VrfMetalGateway schemas by leveraging the `type` enums on VRF and non-VRF IP reservations